src/CoreWCF.Primitives/src/CoreWCF/ServiceAuthorizationManager.cs

```diff
@@ -11,18 +15,52 @@ namespace CoreWCF
 {
     public class ServiceAuthorizationManager
     {
+        private bool _isCheckAccessCoreAsyncOverridden;
+        private bool _isCheckAccessAsyncWithSingleParameterOverridden;
+
+        public ServiceAuthorizationManager()
+        {
+            Type implementorType = GetType();
+            var methods = implementorType.GetMethods(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
+
+            var checkAccessCoreAsyncMethodInfo = methods.Single(static x => x.Name == nameof(CheckAccessCoreAsync));
+            var baseCheckAccessCoreAsyncMethodInfo = checkAccessCoreAsyncMethodInfo!.GetBaseDefinition();
+
+            _isCheckAccessCoreAsyncOverridden = baseCheckAccessCoreAsyncMethodInfo.DeclaringType != checkAccessCoreAsyncMethodInfo.DeclaringType;
+
+            var checkAccessAsyncWithSingleParameterMethodInfo = methods.SingleOrDefault(static x =>
+                x.Name == nameof(CheckAccessAsync) && x.GetParameters().Length == 1);
+            var baseCheckAccessAsyncWithSingleParameterMethodInfo =
+                checkAccessAsyncWithSingleParameterMethodInfo!.GetBaseDefinition();
+
+            _isCheckAccessAsyncWithSingleParameterOverridden =
+                baseCheckAccessAsyncWithSingleParameterMethodInfo.DeclaringType != checkAccessAsyncWithSingleParameterMethodInfo.DeclaringType;
+        }
+
         // This is the API called by framework to perform CheckAccess.
         // The API is responsible for ...
         // 1) Evaluate all policies (Forward\Backward)
-        // 2) Optionally wire up the resulting AuthorizationContext 
+        // 2) Optionally wire up the resulting AuthorizationContext
         //    to ServiceSecurityContext.
-        // 3) An availability of message content to make an authoritive decision. 
+        // 3) An availability of message content to make an authoritive decision.
         // 4) Return the authoritive decision true/false (allow/deny).
-        public virtual bool CheckAccess(OperationContext operationContext, ref Message message)
+        [Obsolete("Implementers should override CheckAccessAsync.")]
+        public virtual bool CheckAccess(OperationContext operationContext, ref Message message) => CheckAccess(operationContext);
+
+        public virtual async ValueTask<(bool isAuthorized, Message message)> CheckAccessAsync(OperationContext operationContext, Message message)
         {
-            return CheckAccess(operationContext);
+            if (_isCheckAccessAsyncWithSingleParameterOverridden || _isCheckAccessCoreAsyncOverridden)
+            {
```

Comment:
Should we combine these two into a single field with the two values combined already? I couldn't see anywhere that either is used independently.

Reply:
I merged both boolean values in a single field